pkg/resources/management.cattle.io/v3/project/mutator.go

```
@@ -85,13 +96,81 @@ func (m *Mutator) Admit(request *admission.Request) (*admissionv1.AdmissionRespo
+		project = m.updateProjectNamespace(project)
 	default:
 		return nil, fmt.Errorf("operation type %q not handled", request.Operation)
 	}
+	response := &admissionv1.AdmissionResponse{}
+	if err := patch.CreatePatch(request.Object.Raw, project, response); err != nil {
+		return nil, fmt.Errorf("failed to create patch: %w", err)
+	}
+	response.Allowed = true
+	return response, nil
+}
+
+func (m *Mutator) createProjectNamespace(project *v3.Project) (*v3.Project, error) {
+	newProject := project.DeepCopy()
+	backingNamespace := ""
+	var err error
+	// When the project name is empty, that means we want to generate a name for it
+	// Name generation happens after mutating webhooks, so in order to have access to the name early
+	// for the backing namespace, we need to generate it ourselves
+	if project.Name == "" {
+		// If err is nil, (meaning "project exists", see below) we need to repeat the generation process to find a project name and backing namespace that isn't taken
+		for err == nil {
+			newName := names.SimpleNameGenerator.GenerateName(project.GenerateName)
+			_, err = m.projectClient.Get(newProject.Spec.ClusterName, newName, v1.GetOptions{})
+			if err == nil {
+				// A project with this name already exists. Generate a new name.
+				continue
+			} else if !apierrors.IsNotFound(err) {
+				return nil, err
+			}
+
+			backingNamespace = name.SafeConcatName(newProject.Spec.ClusterName, strings.ToLower(newName))
+			_, err = m.namespaceClient.Get(backingNamespace, v1.GetOptions{})
+
+			// If the backing namespace already exists, generate a new project name
+			if err != nil && !apierrors.IsNotFound(err) {
+				return nil, err
+			}
+		}
+	} else {
+		backingNamespace = name.SafeConcatName(newProject.Spec.ClusterName, strings.ToLower(newProject.Name))
+		_, err = m.namespaceClient.Get(backingNamespace, v1.GetOptions{})
+		if err == nil {
+			return nil, fmt.Errorf("failed to create project: namespace %v already exists", backingNamespace)
```

```suggestion
			return nil, fmt.Errorf("failed to create project: namespace %s already exists", backingNamespace)
```

---

Q: Should we throw an error or just make it a no-op? 🤔